lighthouse-core/gather/gatherers/dobetterweb/response-compression.js

```diff
@@ -0,0 +1,70 @@
+/**
+ * @license
+ * Copyright 2017 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+ /**
+  * @fileoverview Determines optimized gzip/br/deflate filesizes for all responses by
+  *   checking the content-encoding header.
+  */
+'use strict';
+
+const Gatherer = require('../gatherer');
+
+const compressionTypes = ['gzip', 'br', 'deflate'];
+
+class ResponseCompression extends Gatherer {
+  /**
+   * @param {!NetworkRecords} networkRecords
+   * @return {!Array<{url: string, isBase64DataUri: boolean, mimeType: string, resourceSize: number}>}
+   */
+  static filterUnoptimizedResponses(networkRecords) {
+    return networkRecords.reduce((prev, record) => {
+      const isTextBasedResource = record._resourceType && record._resourceType._isTextType;
+      const isContentEncoded = isTextBasedResource &&
+        record._resourceSize &&
+        !record._responseHeaders.find(header =>
+          header.name.toLowerCase() === 'content-encoding' &&
+          compressionTypes.includes(header.value)
+        );
+
+      if (isTextBasedResource && isContentEncoded) {
+        prev.push({
+          url: record._url,
+          mimeType: record._mimeType,
+          resourceSize: record._resourceSize,
+        });
+      }
+
+      return prev;
+    }, []);
+  }
+
+  afterPass(options, traceData) {
+    const networkRecords = traceData.networkRecords;
+    const textRecords = ResponseCompression.filterUnoptimizedResponses(networkRecords);
+
+    return textRecords.map(record => {
+      // GZIP saving is predictable (@see https://cs.chromium.org/chromium/src/third_party/WebKit/Source/devtools/front_end/audits/AuditRules.js?q=f:AuditRules+gzip&sq=package:chromium&l=97)
+      const gzipSize = record.resourceSize * 2 / 3;
```

Comment:
This is awesome! I'd love to be able to report more accurate estimates based on the makeup of the file types although I'm not sure I trust the methodology behind these numbers and think we'll end up underreporting savings (aside: the current AuditRules.js also says 66% savings not 33%).

27% for script sounds really low especially when you look around at the compression ratios for various libraries and Ilya's table in the read more link.

Specifically, I distrust this assumption.
> If we assume that the files >500 bytes are evenly distributed between compressed/uncompressed

I would *hope*, though this may very naive, that larger, more sophisticated applications with lots of script use GZIP at a higher rate than that of smaller applications with less script. (example: new developer copy/pastes the CDN tag for jQuery/Angular/etc which will be gzipped and then writes a couple of his own files not gzipped at 1/10th the size of the library, which leads to the incorrect conclusion that gzip *increases* file size). Imprecise methodology still is better than no methodology at all though, just wanted to flag it as needing further investigation.